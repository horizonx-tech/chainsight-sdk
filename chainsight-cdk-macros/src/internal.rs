use proc_macro2::Span;
use quote::quote;
use syn::{Ident, Type};

pub fn attrs_query_func() -> proc_macro2::TokenStream {
    quote! {
        #[ic_cdk::query]
        #[candid::candid_method(query)]
    }
}

pub fn attrs_update_func() -> proc_macro2::TokenStream {
    quote! {
        #[ic_cdk::update]
        #[candid::candid_method(update)]
    }
}

pub fn gen_func_quote_to_call_proxy(
    func_name: &str,
    response_type: Type,
    args_type: Option<Type>,
    func_name_to_call: &str,
) -> proc_macro2::TokenStream {
    let func_name_ident = Ident::new(func_name, Span::call_site());
    let func_name_to_call_ident = Ident::new(func_name_to_call, Span::call_site());

    let receiver_provider_quote = if let Some(args_type) = args_type {
        quote! { ReceiverProvider::<#args_type, #response_type> }
    } else {
        quote! { ReceiverProviderWithoutArgs::<#response_type> }
    };

    quote! {
        async fn #func_name_ident(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
            use chainsight_cdk::rpc::Receiver;
            chainsight_cdk::rpc::#receiver_provider_quote::new(
                proxy(),
                #func_name_to_call_ident
            )
            .reply(input)
            .await
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_gen_func_quote_to_call_proxy() {
        let actual = gen_func_quote_to_call_proxy(
            "proxy_get_last_timestamp",
            syn::parse_quote! { u64 },
            None,
            "_get_last_timestamp",
        );
        let expected = quote! {
            async fn proxy_get_last_timestamp(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
                use chainsight_cdk::rpc::Receiver;
                chainsight_cdk::rpc::ReceiverProviderWithoutArgs::<u64>::new(proxy(), _get_last_timestamp)
                    .reply(input)
                    .await
            }
        };

        assert_eq!(actual.to_string(), expected.to_string());
    }
}
