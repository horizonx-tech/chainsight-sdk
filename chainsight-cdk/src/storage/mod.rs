mod storable;
#[warn(clippy::module_inception)]
mod storage;
mod token;
pub use storable::*;
pub use storage::*;
pub use token::*;
