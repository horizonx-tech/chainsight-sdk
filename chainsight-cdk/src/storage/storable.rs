use std::borrow::Cow;

use ic_stable_structures::{BoundedStorable, Storable};

#[derive(
    Debug,
    Clone,
    Default,
    PartialEq,
    PartialOrd,
    candid::CandidType,
    candid::Deserialize,
    serde::Serialize,
)]
pub struct StorableBool(pub bool);

impl Storable for StorableBool {
    fn to_bytes(&self) -> std::borrow::Cow<[u8]> {
        let num: u8 = if self.0 { 1 } else { 0 };
        Cow::Owned(num.to_be_bytes().to_vec())
    }

    fn from_bytes(bytes: std::borrow::Cow<[u8]>) -> Self {
        let num = u8::from_be_bytes(bytes.as_ref().try_into().unwrap());
        match num {
            0 => StorableBool(false),
            1 => StorableBool(true),
            _ => panic!("Invalid bool encoding: expected 0 or 1, found {}", num),
        }
    }
}
impl BoundedStorable for StorableBool {
    const MAX_SIZE: u32 = 1;
    const IS_FIXED_SIZE: bool = true;
}
impl From<StorableBool> for bool {
    fn from(wrapper: StorableBool) -> Self {
        wrapper.0
    }
}
impl From<bool> for StorableBool {
    fn from(unwrapped: bool) -> Self {
        StorableBool(unwrapped)
    }
}

#[derive(
    Debug,
    Clone,
    Default,
    PartialEq,
    PartialOrd,
    candid::CandidType,
    candid::Deserialize,
    serde::Serialize,
)]
pub struct StorableStrings(pub Vec<String>);

impl Storable for StorableStrings {
    fn to_bytes(&self) -> std::borrow::Cow<[u8]> {
        let mut result = Vec::new();

        for s in &self.0 {
            let s_bytes = s.as_bytes();
            let s_len = s_bytes.len() as u64;
            let s_len_bytes = s_len.to_be_bytes().to_vec();

            result.extend_from_slice(&s_len_bytes);
            result.extend_from_slice(s_bytes);
        }

        Cow::Owned(result)
    }

    fn from_bytes(bytes: std::borrow::Cow<[u8]>) -> Self {
        let mut result = Vec::new();
        let mut bytes = bytes.as_ref();

        while !bytes.is_empty() {
            let s_len_bytes = &bytes[0..8];
            let s_len = u64::from_be_bytes(s_len_bytes.try_into().unwrap()) as usize;
            let s_bytes = &bytes[8..(8 + s_len)];
            let s = String::from_utf8(s_bytes.to_vec()).unwrap();

            result.push(s);

            bytes = &bytes[(8 + s_len)..];
        }

        Self(result)
    }
}

impl BoundedStorable for StorableStrings {
    const MAX_SIZE: u32 = 8192;
    const IS_FIXED_SIZE: bool = false;
}

impl From<StorableStrings> for Vec<String> {
    fn from(wrapper: StorableStrings) -> Self {
        wrapper.0
    }
}
impl From<Vec<String>> for StorableStrings {
    fn from(unwrapped: Vec<String>) -> Self {
        StorableStrings(unwrapped)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_storable_strings() {
        let data = vec!["hello".to_string(), "world".to_string()];
        let storable = StorableStrings(data.clone());
        let decoded = StorableStrings::from_bytes(storable.to_bytes());
        assert_eq!(data, decoded.0);

        let data = vec![
            "u3zgx-4yaaa-aaaal-achaa-cai".to_string(),
            "ua42s-gaaaa-aaaal-achcq-cai".to_string(),
            "uh54g-lyaaa-aaaal-achca-cai".to_string(),
            "7fpuj-hqaaa-aaaal-acg7q-cai".to_string(),
        ];
        let storable = StorableStrings(data.clone());
        let decoded = StorableStrings::from_bytes(storable.to_bytes());
        assert_eq!(data, decoded.0);
    }
}
