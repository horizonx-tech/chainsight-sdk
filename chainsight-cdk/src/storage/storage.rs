use std::{borrow::Cow, cell::RefCell, collections::HashMap};

use candid::{CandidType, Decode, Encode};
use ic_stable_structures::memory_manager::{MemoryId, VirtualMemory};
use ic_stable_structures::StableBTreeMap;
use ic_stable_structures::{
    memory_manager::MemoryManager, BoundedStorable, DefaultMemoryImpl, Storable,
};
use serde::Deserialize;

use super::token::Token;
type Memory = VirtualMemory<DefaultMemoryImpl>;

#[derive(Deserialize, CandidType, Clone)]
pub struct Values(Vec<Data>);
#[derive(Deserialize, CandidType, Clone)]

pub struct Data(HashMap<String, Token>);

impl Data {
    pub fn new(val: HashMap<String, Token>) -> Self {
        Self(val)
    }
}

impl Storable for Values {
    fn to_bytes(&self) -> std::borrow::Cow<[u8]> {
        Cow::Owned(Encode!(self).unwrap())
    }

    fn from_bytes(bytes: std::borrow::Cow<[u8]>) -> Self {
        Decode!(bytes.as_ref(), Self).unwrap()
    }
}
impl BoundedStorable for Values {
    const MAX_SIZE: u32 = 100_000;
    const IS_FIXED_SIZE: bool = false;
}

impl Values {
    fn new() -> Self {
        Self(Vec::new())
    }
    fn append(&mut self, data: Data) {
        self.0.push(data);
    }
    pub fn to_vec(&self) -> Vec<Data> {
        self.0.clone()
    }
}

impl Data {
    pub fn get(&self, key: &str) -> Option<&Token> {
        self.0.get(key)
    }
}

thread_local! {
    static MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> = RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
    static MAP: RefCell<StableBTreeMap<u64, Values, Memory>> = RefCell::new(
        StableBTreeMap::init(
            MANAGER.with(|m|m.borrow().get(MemoryId::new(0))),
        )
    )
}

pub fn insert(key: u64, data: Data) {
    MAP.with(|m| {
        m.borrow_mut()
            .get(&key)
            .get_or_insert(Values::new())
            .append(data)
    })
}

pub fn between(from: u64, to: u64) -> Vec<(u64, Values)> {
    MAP.with(|m| {
        m.borrow()
            .range(from..to)
            .map(|(k, v)| (k, v.clone()))
            .collect()
    })
}

// get last n
pub fn last(n: u64) -> Vec<(u64, Values)> {
    let length = MAP.with(|m| m.borrow().len());
    if length <= n {
        MAP.with(|m| m.borrow().iter().map(|(k, v)| (k, v.clone())).collect())
    } else {
        MAP.with(
            |m: &RefCell<
                StableBTreeMap<u64, Values, VirtualMemory<std::rc::Rc<RefCell<Vec<u8>>>>>,
            >| {
                m.borrow()
                    .iter()
                    .skip((length - n) as usize)
                    .map(|(k, v)| (k, v.clone()))
                    .collect()
            },
        )
    }
}
