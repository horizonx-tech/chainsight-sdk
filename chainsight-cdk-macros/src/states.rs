use chainsight_cdk::storage::Data;
use proc_macro::TokenStream;
use quote::{quote, ToTokens};
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, parse_quote, Expr, LitBool, LitStr, Type,
};

use crate::internal::{attrs_query_func, attrs_update_func, gen_func_quote_to_call_proxy};

pub trait Persist {
    fn untokenize(data: Data) -> Self;
    fn tokenize(&self) -> Data;
}

pub fn persist_derive(input: TokenStream) -> TokenStream {
    let input = syn::parse_macro_input!(input as syn::DeriveInput);
    let name = input.ident;
    let fields = match input.data {
        syn::Data::Struct(syn::DataStruct {
            fields: syn::Fields::Named(syn::FieldsNamed { ref named, .. }),
            ..
        }) => named,
        _ => panic!("Only support struct"),
    };
    // get field name and type
    let mut field_names = Vec::new();
    let mut field_types = Vec::new();
    let mut untokenize_functions = Vec::new();
    // let mut tokenize_functions = Vec::new();
    for field in fields {
        let field_name = field.ident.as_ref().unwrap();
        let field_type = &field.ty;
        field_names.push(field_name);
        field_types.push(field_type);
        let untokenize_function = match field_type.to_token_stream().to_string().as_str() {
            "String" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_string(),
            },
            "u128" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_u128().unwrap(),
            },
            "u64" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_u64().unwrap(),
            },
            "u32" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_u32().unwrap(),
            },
            "u16" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_u16().unwrap(),
            },
            "u8" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_u8().unwrap(),
            },
            "i128" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_i128().unwrap(),
            },
            "i64" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_i64().unwrap(),
            },
            "i32" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_i32().unwrap(),
            },
            "i16" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_i16().unwrap(),
            },
            "i8" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_i8().unwrap(),
            },
            "bool" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_bool().unwrap(),
            },
            "Vec<u8>" => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_vec().unwrap(),
            },
            "chainsight_cdk::core::U256" => quote! {
               #field_name: data.get(stringify!(#field_name)).unwrap().to_u256().unwrap(),
            },
            "U256" => quote! {
               #field_name: data.get(stringify!(#field_name)).unwrap().to_u256().unwrap(),
            },
            _ => quote! {
                #field_name: data.get(stringify!(#field_name)).unwrap().to_string(),
            },
        };
        untokenize_functions.push(untokenize_function);
    }
    let expanded = quote! {
        impl #name {
            pub fn _tokenize(&self) -> chainsight_cdk::storage::Data {
                let mut data: HashMap<std::string::String, chainsight_cdk::storage::Token> = HashMap::new();
                #(
                    data.insert(stringify!(#field_names).to_string(), chainsight_cdk::storage::Token::from(self.#field_names.clone()));
                )*
                Data::new(data)
            }
            pub fn _untokenize(data: Data) -> Self {
                Self {
                    #(#untokenize_functions)*
                }
            }
        }
        impl chainsight_cdk::storage::Persist for #name {
            fn untokenize(data: Data) -> Self {
                Self::_untokenize(data)
            }
            fn tokenize(&self) -> Data {
                self._tokenize()
            }
        }
    };

    expanded.into()
}

struct SingleStateInput {
    name: LitStr,
    ty: Type,
    is_expose_getter: LitBool,
    init: Option<Expr>,
}
impl Parse for SingleStateInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name: LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let ty: Type = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let is_expose_getter: LitBool = input.parse()?;
        let init = if input.peek(syn::Token![,]) {
            input.parse::<syn::Token![,]>()?;
            Some(input.parse()?)
        } else {
            None
        };
        Ok(SingleStateInput {
            name,
            ty,
            is_expose_getter,
            init,
        })
    }
}
pub fn manage_single_state(input: TokenStream) -> TokenStream {
    let args = parse_macro_input!(input as SingleStateInput);
    manage_single_state_internal(args).into()
}
fn manage_single_state_internal(args: SingleStateInput) -> proc_macro2::TokenStream {
    let SingleStateInput {
        name,
        ty,
        is_expose_getter,
        init,
    } = args;

    let var_ident = syn::Ident::new(&name.value().to_uppercase(), name.span());
    let get_ident = syn::Ident::new(&format!("get_{}", name.value()), name.span());
    let set_ident = syn::Ident::new(&format!("set_{}", name.value()), name.span());

    let init = match init {
        Some(init_value) => quote!(#init_value),
        None => quote!(std::default::Default::default()),
    };
    let getter_derives = if is_expose_getter.value {
        attrs_query_func()
    } else {
        quote! {}
    };

    quote! {
        thread_local! {
            static #var_ident: std::cell::RefCell<#ty> = std::cell::RefCell::new(#init);
        }

        #getter_derives
        pub fn #get_ident() -> #ty {
            #var_ident.with(|state| state.borrow().clone())
        }

        pub fn #set_ident(value: #ty) {
            #var_ident.with(|state| *state.borrow_mut() = value);
        }
    }
}

struct VecStateInput {
    name: LitStr,
    ty: Type,
    is_expose_getter: LitBool,
}
impl Parse for VecStateInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let ty = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let is_expose_getter: LitBool = input.parse()?;
        Ok(VecStateInput {
            name,
            ty,
            is_expose_getter,
        })
    }
}
pub fn manage_vec_state(input: TokenStream) -> TokenStream {
    let args = parse_macro_input!(input as VecStateInput);
    manage_vec_state_internal(args).into()
}
fn manage_vec_state_internal(args: VecStateInput) -> proc_macro2::TokenStream {
    let VecStateInput {
        name,
        ty,
        is_expose_getter,
    } = args;

    let state_name = name.value();
    let state_upper_name = syn::Ident::new(&format!("{}S", state_name.to_uppercase()), name.span());
    let get_vec_func = syn::Ident::new(&format!("get_{}s", state_name), name.span());
    let _get_vec_func_str = format!("_get_{}s", state_name);
    let _get_vec_func = syn::Ident::new(&_get_vec_func_str, name.span());
    let proxy_get_vec_func_quote = gen_func_quote_to_call_proxy(
        &format!("proxy_get_{}s", state_name),
        parse_quote! { Vec<#ty> },
        None,
        &_get_vec_func_str,
    );
    let get_len_func = syn::Ident::new(&format!("{}s_len", state_name), name.span());
    let _get_len_func_str = format!("_{}s_len", state_name);
    let _get_len_func = syn::Ident::new(&_get_len_func_str, name.span());
    let proxy_get_len_func_quote = gen_func_quote_to_call_proxy(
        &format!("proxy_{}s_len", state_name),
        parse_quote! { usize },
        None,
        &_get_len_func_str,
    );
    let get_last_elem_func = syn::Ident::new(&format!("get_last_{}", state_name), name.span());
    let _get_last_elem_func_str = format!("_get_last_{}", state_name);
    let _get_last_elem_func = syn::Ident::new(&_get_last_elem_func_str, name.span());
    let proxy_get_last_elem_func_quote = gen_func_quote_to_call_proxy(
        &format!("proxy_get_last_{}", state_name),
        parse_quote! { #ty },
        None,
        &_get_last_elem_func_str,
    );
    let get_top_elems_func = syn::Ident::new(&format!("get_top_{}s", state_name), name.span());
    let _get_top_elems_func_str = format!("_get_top_{}s", state_name);
    let _get_top_elems_func = syn::Ident::new(&_get_top_elems_func_str, name.span());
    let proxy_get_top_elems_func_quote = gen_func_quote_to_call_proxy(
        &format!("proxy_get_top_{}s", state_name),
        parse_quote! { Vec<#ty> },
        Some(parse_quote! { usize }),
        &_get_top_elems_func_str,
    );
    let get_elem_func = syn::Ident::new(&format!("get_{}", state_name), name.span());
    let _get_elem_func_str = format!("_get_{}", state_name);
    let _get_elem_func = syn::Ident::new(&_get_elem_func_str, name.span());
    let proxy_get_elem_func_quote = gen_func_quote_to_call_proxy(
        &format!("proxy_get_{}", state_name),
        parse_quote! { #ty },
        Some(parse_quote! { usize }),
        &_get_elem_func_str,
    );
    let add_elem_func = syn::Ident::new(&format!("add_{}", state_name), name.span());

    let getter_derives = if is_expose_getter.value {
        attrs_query_func()
    } else {
        quote! {}
    };
    let update_derive = attrs_update_func();

    quote! {
        thread_local! {
            static #state_upper_name: std::cell::RefCell<Vec<#ty>> = std::cell::RefCell::new(Vec::new());
        }

        #getter_derives
        fn #get_vec_func() -> Vec<#ty> {
            #_get_vec_func()
        }

        pub fn #_get_vec_func() -> Vec<#ty> {
            #state_upper_name.with(|state| state.borrow().clone())
        }

        #update_derive
        #proxy_get_vec_func_quote

        #getter_derives
        fn #get_len_func() -> usize {
            #_get_len_func()
        }

        pub fn #_get_len_func() -> usize {
            #state_upper_name.with(|state| state.borrow().len())
        }

        #update_derive
        #proxy_get_len_func_quote

        #getter_derives
        fn #get_last_elem_func() -> #ty {
            #_get_last_elem_func()
        }

        pub fn #_get_last_elem_func() -> #ty {
            #state_upper_name.with(|state| state.borrow().last().unwrap().clone())
        }

        #update_derive
        #proxy_get_last_elem_func_quote


        #getter_derives
        fn #get_top_elems_func(n: usize) -> Vec<#ty> {
            #_get_top_elems_func(n)
        }

        pub fn #_get_top_elems_func(n: usize) -> Vec<#ty> {
            #state_upper_name.with(|state| state.borrow().iter().rev().take(n).cloned().collect::<Vec<_>>())
        }

        #update_derive
        #proxy_get_top_elems_func_quote

        #getter_derives
        fn #get_elem_func(idx: usize) -> #ty {
            #_get_elem_func(idx)
        }

        pub fn #_get_elem_func(idx: usize) -> #ty {
            #state_upper_name.with(|state| state.borrow()[idx].clone())
        }

        #update_derive
        #proxy_get_elem_func_quote

        pub fn #add_elem_func(value: #ty) {
            #state_upper_name.with(|state| state.borrow_mut().push(value));
        }
    }
}

struct MapStateInput {
    name: LitStr,
    key_ty: Type,
    val_ty: Type,
    is_expose_getter: LitBool,
}
impl Parse for MapStateInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let key_ty = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let val_ty = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let is_expose_getter: LitBool = input.parse()?;
        Ok(MapStateInput {
            name,
            key_ty,
            val_ty,
            is_expose_getter,
        })
    }
}
pub fn manage_map_state(input: TokenStream) -> TokenStream {
    let args = parse_macro_input!(input as MapStateInput);
    manage_map_state_internal(args).into()
}
// todo: support functions to call proxy
fn manage_map_state_internal(args: MapStateInput) -> proc_macro2::TokenStream {
    let MapStateInput {
        name,
        key_ty,
        val_ty,
        is_expose_getter,
    } = args;

    let state_name = name.value();
    let state_upper_name = syn::Ident::new(&format!("{}S", state_name.to_uppercase()), name.span());
    let len_func = syn::Ident::new(&format!("{}s_len", state_name), name.span());
    let get_elem_func = syn::Ident::new(&format!("get_{}", state_name), name.span());
    let insert_elem_func = syn::Ident::new(&format!("insert_{}", state_name), name.span());

    let getter_derives = if is_expose_getter.value {
        attrs_query_func()
    } else {
        quote! {}
    };

    quote! {
        thread_local! {
            static #state_upper_name: std::cell::RefCell<std::collections::HashMap<#key_ty, #val_ty>> = std::cell::RefCell::new(std::collections::HashMap::new());
        }

        #getter_derives
        pub fn #len_func() -> usize {
            #state_upper_name.with(|state| state.borrow().len())
        }

        #getter_derives
        pub fn #get_elem_func(key: #key_ty) -> #val_ty {
            #state_upper_name.with(|state| state.borrow().get(&key).cloned().expect("key not found"))
        }

        pub fn #insert_elem_func(key: #key_ty, value: #val_ty) {
            #state_upper_name.with(|state| state.borrow_mut().insert(key, value));
        }
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot_manage_single_state() {
        let input = quote! {"timer_id", TimerId, true};
        let args: syn::Result<SingleStateInput> = syn::parse2(input);
        let generated = manage_single_state_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__manage_single_state", formatted);
    }

    #[test]
    fn test_snapshot_manage_vec_state() {
        let input = quote! {"snapshots", Snapshot, true};
        let args: syn::Result<VecStateInput> = syn::parse2(input);
        let generated = manage_vec_state_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__manage_vec_state", formatted);
    }

    #[test]
    fn test_snapshot_manage_map_state() {
        let input = quote! {"time_differences", String, u64, true};
        let args: syn::Result<MapStateInput> = syn::parse2(input);
        let generated = manage_map_state_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__manage_map_state", formatted);
    }
}
