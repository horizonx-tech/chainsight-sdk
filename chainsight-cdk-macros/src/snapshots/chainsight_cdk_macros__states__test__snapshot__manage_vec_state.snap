---
source: chainsight-cdk-macros/src/states.rs
expression: formatted
---
thread_local! { static SNAPSHOTSS : std :: cell :: RefCell < Vec < Snapshot >> = std :: cell :: RefCell :: new (Vec :: new ()) ; }
#[ic_cdk::query]
#[candid::candid_method(query)]
fn get_snapshotss() -> Vec<Snapshot> {
    _get_snapshotss()
}
pub fn _get_snapshotss() -> Vec<Snapshot> {
    SNAPSHOTSS.with(|state| state.borrow().clone())
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_get_snapshotss(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProviderWithoutArgs::<Vec<Snapshot>>::new(proxy(), _get_snapshotss)
        .reply(input)
        .await
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn snapshotss_len() -> usize {
    _snapshotss_len()
}
pub fn _snapshotss_len() -> usize {
    SNAPSHOTSS.with(|state| state.borrow().len())
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_snapshotss_len(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProviderWithoutArgs::<usize>::new(proxy(), _snapshotss_len)
        .reply(input)
        .await
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn get_last_snapshots() -> Snapshot {
    _get_last_snapshots()
}
pub fn _get_last_snapshots() -> Snapshot {
    SNAPSHOTSS.with(|state| state.borrow().last().unwrap().clone())
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_get_last_snapshots(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProviderWithoutArgs::<Snapshot>::new(proxy(), _get_last_snapshots)
        .reply(input)
        .await
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn get_top_snapshotss(n: usize) -> Vec<Snapshot> {
    _get_top_snapshotss(n)
}
pub fn _get_top_snapshotss(n: usize) -> Vec<Snapshot> {
    SNAPSHOTSS.with(|state| {
        state
            .borrow()
            .iter()
            .rev()
            .take(n)
            .cloned()
            .collect::<Vec<_>>()
    })
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_get_top_snapshotss(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProvider::<usize, Vec<Snapshot>>::new(proxy(), _get_top_snapshotss)
        .reply(input)
        .await
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn get_snapshots(idx: usize) -> Snapshot {
    _get_snapshots(idx)
}
pub fn _get_snapshots(idx: usize) -> Snapshot {
    SNAPSHOTSS.with(|state| state.borrow()[idx].clone())
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_get_snapshots(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProvider::<usize, Snapshot>::new(proxy(), _get_snapshots)
        .reply(input)
        .await
}
pub fn add_snapshots(value: Snapshot) {
    SNAPSHOTSS.with(|state| state.borrow_mut().push(value));
}

